### DISCUSSION TITLE
Applying semantics multiple times to different inputs

### DISCUSSION THREAD

**Discussion by samguyer** [ohmjs/ohm#469] [Help]
Hello! I am new to Ohm, but have spent many decades working with parsers in various languages. (Compared to tools like yacc/lex, Ohm is such a joy to use!) What I'm trying to do is create a simple query language, which I will use to filter a data set. The query language is essentially just a predicate that will be applied to each data item. What I want to do is parse the filter expression, attach the semantics, but then apply those semantics to each data item to determine if it should be included in the results or not. I don't see an obvious way to invoke the semantic action with an argument. I suppose I could just use closure capture to allow the semantic functions to directly access the data set, but that seems less appealing from a design standpoint. Is there a better way?
Thanks!

**Comment by millergarym** [comment]
@samguyer the `addOperation` can be specified with a signature.
The semantic action can then be invoked passing arguments.
Please see
https://ohmjs.org/docs/api-reference#semantics-objects

For a working example see 
https://github.com/ohmjs/ohm/blob/main/packages/ohm-js/extras/semantics-toAST.js#L31
